Add meta and OpInfo for _embedding_bag_dense_backward

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109211

The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

This is exercised in vdd, internal test `buck2 run '@fbcode//mode/opt' fbcode//pytorch/benchmark/fb/test_gpu:run_test_gpu -- 'pytorch.benchmark.fb.test_gpu.test_gpu.TestBenchmarkFbGpu.test_train_blue_reels_vdd_v3_inductor_speedup'`

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Add meta and OpInfo for _embedding_bag_dense_backward

The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add meta and OpInfo for _embedding_bag_dense_backward"


The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

This is exercised in vdd, internal test `buck2 run 'fbcode//mode/opt' fbcode//pytorch/benchmark/fb/test_gpu:run_test_gpu -- 'pytorch.benchmark.fb.test_gpu.test_gpu.TestBenchmarkFbGpu.test_train_blue_reels_vdd_v3_inductor_speedup'`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7290,6 +7290,69 @@ def sample_inputs_dropout_backward(op_info, device, dtype, requires_grad, **kwar
+        sample.kwargs.pop('max_norm', None)  # not used by the op
+        sample.kwargs.pop('norm_type', None)  # not used by the op
+
+        # This is a precondition for the function, so force it
```
Isn't this already True if it is a precondition for the forward?

ezyang replied:
No, not precondition for forward. The backward only gets called in contexts where the internal private function was guaranteed to be contiguous

albanD replied:
Ok ok

albanD reviewed (approved):
SGTM

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Add meta and OpInfo for _embedding_bag_dense_backward"


The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

This is exercised in vdd, internal test `buck2 run 'fbcode//mode/opt' fbcode//pytorch/benchmark/fb/test_gpu:run_test_gpu -- 'pytorch.benchmark.fb.test_gpu.test_gpu.TestBenchmarkFbGpu.test_train_blue_reels_vdd_v3_inductor_speedup'`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -r

ezyang commented:
@pytorchbot merge -r main

ezyang committed:
Update on "Add meta and OpInfo for _embedding_bag_dense_backward"


The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

This is exercised in vdd, internal test `buck2 run 'fbcode//mode/opt' fbcode//pytorch/benchmark/fb/test_gpu:run_test_gpu -- 'pytorch.benchmark.fb.test_gpu.test_gpu.TestBenchmarkFbGpu.test_train_blue_reels_vdd_v3_inductor_speedup'`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add meta and OpInfo for _embedding_bag_dense_backward"


The sample inputs is a bit involved because there are a lot of
shenanigans in the derivative formula.  Check comments.

This is exercised in vdd, internal test `buck2 run 'fbcode//mode/opt' fbcode//pytorch/benchmark/fb/test_gpu:run_test_gpu -- 'pytorch.benchmark.fb.test_gpu.test_gpu.TestBenchmarkFbGpu.test_train_blue_reels_vdd_v3_inductor_speedup'`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "Sorry I think the test_ops.py::TestCommonCUDA::test_compare_cpu__embedding_bag_dense_backward_cuda_float32 is failing https://hud.pytorch.org/pytorch/pytorch/commit/492a93d185a74bb4ccb9dea184a71f43f8952083 https://github.com/pytorch/pytorch/actions/runs/6190707847/job/16808644559 not sure why this is run in slow when it looks to be a new test" -c nosignal
